[WIP] ui interaction

scara82 opened the pull request:
Without having the family-model-branch merged is hard to test if is working properly and if I'm missing something.

At the moment this PR does:
- Add single form and responders gems with the relative files required
- Add families controller with new, create and index methods
- Add the relative routes and views files
- Add testing for GET and POST
- Add features test
- Add translation in italian for yml files

Commit:
Can create new family and view all families

Commit:
Add simple_form gem

Commit:
Add @family at families#new to fix the form non rendering

Commit:
Configure responders

Commit:
Add specs for families#create

apalski commented on spec/controllers/families_controller_spec.rb:
```diff
@@ -0,0 +1,43 @@
+      expect { post :create, params: { family: family_params } }.to change(Family, :count).by(1)
+    end
+  end
+end
```
This is just an information comment:
I've used some shoulda matchers that I found for my routing specs:
it { is_expected.to route(:get, "/users").to(controller: :users, action: :index) }
I am interested to hear Elle's opinion on their use - it made for smaller specs!

scara82 replied:
Here I was tryint to test if a new family is created, your line is testing instead familes#index?! Not sure if I get you rcomment right.. 🤔

apalski replied:
Sorry, my users#index spec is just an example of a different and shorter way I found to do the routing specs, I have put it in my most recent PR to get Elle's opinion. I wasn't suggesting you change anything, it was just to show it to you.

scara82 replied:
Ahhh, sorry, of course I didn't understand! 😅 Once I will be able to test against the model I will try it, thanks!

elle replied:
I would usually have:
1. feature specs for seeing all families, and creating a new family (happy path)
2. controller specs for the sad path - i.e. trying to create a new family without required attributes => testing to see that the flash message was set correctly, and the rendered template is `:new` rather than `:index`, and that the object has error messages on it

scara82 replied:
I created the folder for features specks and tested for add a family and vie all the family. I left in the controller specs the sad path.

Thanks!

apalski commented on config/locales/simple_form.en.yml:
```diff
@@ -0,0 +1,12 @@
+      # When using html, text and mark won't be used.
+      # html: '<abbr title="required">*</abbr>'
+      error_notification:
+        default_message: "Please review the problems below:"
```
Are you doing Italian versions of these two .yml files?

scara82 replied:
Yes, I will, I'm waiting to have everything working properly for english first. (even if, actually, this shouldn't change so it should be ok adding the italian version)! Thanks for the hint!

apalski commented on config/locales/responders.en.yml:
```diff
@@ -0,0 +1,12 @@
+         alert: '%{resource_name} could not be updated.'
+       destroy:
+         notice: '%{resource_name} was successfully destroyed.'
+         alert: '%{resource_name} could not be destroyed.'
```
I'm not sure whether you should have included the destroy notice and alert yet as you don't yet have it in your controller and routes?

scara82 replied:
The file has been created automatically with the responders inplementation. I will need this so I think there is not point to delete it to add it again late. I think! ;) But good point!

apalski replied:
No, sorry didn't realise that, I would leave it there too!

apalski commented on app/views/families/new.html.erb:
```diff
@@ -0,0 +1,13 @@
+  <%= f.input :bus_driver, as: :radio_buttons, collection: ['Y', 'N'] %>
+  <%= f.input :active, as: :radio_buttons, collection: ['Y', 'N'] %>
+  <%= f.button :submit %>
+<% end %>
```
Does Add a Family need to use translations? I haven't gotten to views yet, so I'm just wondering.

scara82 replied:
Yes, it will need, I will once everyhting is working properly! Thanks!

elle replied:
Yes, and also the field placeholders. 

For the fields labels and placeholders, check `simple_form` documentation about working with translations

scara82 replied:
Done, thanks!

elle commented on app/controllers/families_controller.rb:
```diff
@@ -0,0 +1,30 @@
+
+  def index
+    @families = Family.all
+  end
```
A suggestion to move `index` action to be the first in the controller, because convention

scara82 replied:
Done! Thanks I didn't know about that convention.

elle commented on app/controllers/families_controller.rb:
```diff
@@ -0,0 +1,30 @@
+  end
+
+  private
+  def family_params
```
Space between lines 16 and 17

scara82 replied:
Done. Thanks!

elle commented on app/views/families/index.html.erb:
```diff
@@ -0,0 +1,4 @@
+<a href="families/new">Add a Family</a>
+<p>List of Families:</p>
+<%= render @families %>
+<br>
```
Is the `<br>` needed here? or could you use css to give it more space?

scara82 replied:
I added it just for the moment because it looks nicer, but I will fix it properly once strating with css. Thanks!

elle commented on app/views/families/new.html.erb:
```diff
@@ -0,0 +1,13 @@
+  <%= f.input :preferred_gender, as: :radio_buttons, collection: ['F', 'M'] %>
+  <%= f.input :bus_driver, as: :radio_buttons, collection: ['Y', 'N'] %>
+  <%= f.input :active, as: :radio_buttons, collection: ['Y', 'N'] %>
+  <%= f.button :submit %>
```
Double quotes

scara82 replied:
Done. Thanks!

elle commented on app/views/layouts/application.html.erb:
```diff
@@ -9,6 +9,7 @@
   </head>
 
```
Extra newline

scara82 replied:
Done. Thanks!

elle commented on config/application.rb:
```diff
@@ -15,19 +15,20 @@
-    generate.test_framework :rspec
-    generate.view_specs false
+    config.generators do |generate|
+      generate.helper false
```
I usually leave this one as the default true, but up to you

scara82 replied:
So probably it would be better, I changed it, thanks! :)

elle commented on config/locales/responders.en.yml:
```diff
@@ -0,0 +1,12 @@
+         alert: '%{resource_name} could not be updated.'
+       destroy:
+         notice: '%{resource_name} was successfully destroyed.'
+         alert: '%{resource_name} could not be destroyed.'
```
Double quotes

scara82 replied:
Done. Thanks!

elle commented on config/locales/simple_form.en.yml:
```diff
@@ -0,0 +1,12 @@
+    "no": 'No'
+    required:
+      text: 'required'
+      mark: '*'
```
Double quotes

scara82 replied:
Done. Thanks!

elle commented on config/routes.rb:
```diff
@@ -1,3 +1,5 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  resources :families, only: [:index, :new, :create]
```
👍

elle commented on spec/controllers/families_controller_spec.rb:
```diff
@@ -0,0 +1,43 @@
+require 'rails_helper'
```
Double quotes

scara82 replied:
Done. Thanks!

elle commented on app/assets/javascripts/families.coffee:
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
This file is not really needed

scara82 replied:
Deleted, thanks!

elle commented on app/views/families/index.html.erb:
```diff
@@ -0,0 +1,4 @@
+<a href="families/new">Add a Family</a>
```
Use translations

scara82 replied:
Done, thanks!

elle commented on app/views/families/index.html.erb:
```diff
@@ -0,0 +1,4 @@
+<a href="families/new">Add a Family</a>
+<p>List of Families:</p>
```
Same, translations

scara82 replied:
Done, thanks!

Commit:
Move families#index to the top to follew covention

Commit:
Add location for respond_with and add to config/application.rb responder controller generator to use respond_with

Commit:
Replace families with action in flash message to have it more generic and reusable

Commit:
Add features spec folder and features/families_spec.rb file

Commit:
Populate translations files, en and it, and use it in the families/index.html.erb

Commit:
Emininates the placeholders from the new_family _form

Commit:
Move basic_http rack auth to production config

- Because it was interfering with our feature specs

Commit:
Add double quotes

Commit:
Add features spec for adding a family

Commit:
Add capybara_webkit.rb

Commit:
Add spec/controller folder

Commit:
Add families_controller specs for sad path

Commit:
Add features spec for admin that wants to see all families in the homepage

Commit:
Rubocop fixes

Commit:
deleted unnecessary families.coffee file

Commit:
New family form: user F/M format

- Chage radio buttons in new_family form with default value
  and change f to Female and M to Male fo preferred gender.
- Converts in double quotes the simple_form.en.yml file

Commit:
Add italian traslation for config/locales/it.yml-responders.it.yml-simple_form.it.yml

Commit:
Fix missing double quote
